### PR TITLE
Issue #6473 - canonicalPath refactor & fix alias check in PathResource (Jetty-10)

### PR DESCRIPTION
**Pull request by lachlan-roberts**
## Issue #6473

Merge of PR #6474 and PR #6477 for Jetty-10.

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 1]
```diff
@@ -31,54 +33,94 @@
  * via the static methods such as {@link #build()} and {@link #from(String)}.
  *
  * A URI such as
- * <code>http://user@host:port/path;ignored/info;param?query#ignored</code>
- * is split into the following undecoded elements:<ul>
+ * {@code http://user@host:port/path;param1/%2e/info;param2?query#fragment}
+ * this class will split it into the following optional elements:<ul>
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: The grammar of this sentence is split between old and new version:

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 2]
```diff
@@ -34,7 +34,7 @@
  *
  * A URI such as
  * {@code http://user@host:port/path;param1/%2e/info;param2?query#fragment}
- * this class will split it into the following optional elements:<ul>
+ * are split into the following optional elements:<ul>
```
Comment (gregw) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]:
```suggestion
 * is split into the following optional elements:<ul>
```

**Review by gregw — CHANGES_REQUESTED** [review]
sorry to be pendantic

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 3]
```diff
@@ -31,54 +33,94 @@
...
+ * <p>
+ * If there are multiple path parameters, only the last one is returned by {@link #getParam()}.
+ * </p>
+ **/
 public interface HttpURI
 {
+    /**
+     * Violations of safe URI interpretations
+     */
     enum Violation
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: This class was called `Ambiguous` so this is a breaking change.

Also, why does it not implement `ComplianceViolation`, which seems to summarize the violation concept?

On a closer look, I think this class here is a complete duplicate of `UriCompliance.Violation`.
It's here because in 9.4.x we don't have `UriCompliance` (so it's a merge side-effect).
Should it not be completely removed, so that in 10 we only use `UriCompliance.Violation`?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 4]
```diff
@@ -166,15 +213,26 @@ static Immutable from(String scheme, String host, int port, String pathQuery)
...
+     * @param violation the violation to check.
+     * @return true if the URI has the passed violation.
+     */
+    boolean hasViolation(Violation violation);
+
+    /**
+     * @return Set of violations in the URI.
+     */
+    Collection<Violation> getViolations();
+
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:188]: See comment above, I think methods that expose `Violation` should be removed or changed to use `UriCompliance.Violation`.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 5]
```diff
@@ -1216,6 +1298,10 @@ else if (c == '/')
                             {
                                 switch (encodedValue)
                                 {
+                                    case 0:
+                                    // Byte 0 cannot be present in a UTF-8 sequence in any position
+                                    // other than as the NUL ASCII byte which we do not wish to allow.
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: Fix indentation.

File: jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java [hunk 6]
```diff
@@ -231,6 +227,11 @@ public static UriCompliance from(String spec)
             boolean exclude = element.startsWith("-");
             if (exclude)
                 element = element.substring(1);
+
+            // Ignore removed name. TODO: remove in future release.
+            if (element.equals("NON_CANONICAL_AMBIGUOUS_PATHS"))
+                continue;
+
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:232]: This block should be removed, because the constant `NON_CANONICAL_AMBIGUOUS_PATHS` was only present in PRs, never made into actual releases.

File: jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java [hunk 7]
```diff
@@ -330,4 +331,47 @@ public String toString()
...
+    public static String checkUriCompliance(UriCompliance compliance, HttpURI uri)
+    {
+        for (HttpURI.Violation violation : HttpURI.Violation.values())
+        {
+            if (uri.hasViolation(violation) && (compliance == null || !compliance.allows(__uriViolations.get(violation))))
+                return violation.getMessage();
+        }
+        return null;
+    }
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:341]: See previous comment. This mapping from `HttpURI.Violation` to `UriCompliance.Violation` should not be necessary.
`HttpURI` should just use `UriCompliance.Violation`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java [hunk 8]
```diff
@@ -2305,6 +2308,8 @@ public InputStream getResourceAsStream(String path)
         @Override
         public Set<String> getResourcePaths(String path)
         {
+            if (path == null)
+                return null;
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:2321]: Should not the logic here be similar to `getResource()`? I think we should canonicalize here too.

And also in `getRealPath()`.

I think we need to review the `ServletContext` implementation more carefully.

File: jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java [hunk 9]
```diff
@@ -1713,15 +1713,6 @@ public void testAmbiguousPaths() throws Exception
...
             containsString("pathInfo=/path/info")));
-
-        _connector.getBean(HttpConnectionFactory.class).getHttpConfiguration().setUriCompliance(UriCompliance.from(EnumSet.of(
-            UriCompliance.Violation.AMBIGUOUS_PATH_SEPARATOR,
-            UriCompliance.Violation.AMBIGUOUS_PATH_SEGMENT,
-            UriCompliance.Violation.AMBIGUOUS_PATH_PARAMETER,
-            UriCompliance.Violation.NON_CANONICAL_AMBIGUOUS_PATHS)));
-        assertThat(_connector.getResponse(request), Matchers.allOf(
-            startsWith("HTTP/1.1 200"),
-            containsString("pathInfo=/path/ambiguous/.././info")));
```
Comment (sbordet) [inline jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java:1724]: Seems like a legit test. Why was it removed?

File: jetty-servlet/src/main/java/org/eclipse/jetty/servlet/DefaultServlet.java [hunk 10]
```diff
@@ -420,8 +419,7 @@ else if (_servletContext instanceof ContextHandler.Context)
             }
             else
             {
-                URL u = _servletContext.getResource(pathInContext);
-                r = _contextHandler.newResource(u);
```
Comment (sbordet) [inline jetty-servlet/src/main/java/org/eclipse/jetty/servlet/DefaultServlet.java:424]: Why was this changed? Can `_servletContext` never be other than `ContextHandler.Context`?

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 11]
```diff
@@ -471,6 +471,7 @@ public static String decodePath(String path, int offset, int length)
                             if (u == 'u')
                             {
                                 // UTF16 encoding is only supported with UriCompliance.Violation.UTF16_ENCODINGS.
+                                // This is wrong. This is a codepoint not a char
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:474]: This needs to be fixed then? 9.4.x too?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java (hunk 3 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: Ambiguous was public for a few releases, but not used in any interfaces at all, so I think it is OK to change.
But worth investigating if we can actually use UriCompliance.Violation here....

File: jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java (hunk 6 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:232]: are we sure... if so, then yes

File: jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java (hunk 9 above)
Comment (gregw) [reply jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java:1724]: becaues the compliance mode it was testing no longer exists.

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 11 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:474]: Nope - this is just how we implemented %u.  It is not strictly correct, but let's not improve a feature we really want to remove.

File: jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java (hunk 6 above)
Comment (sbordet) [reply jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:232]: Doh! I was wrong, keep this code, as the constant was in 10.0.5.
